pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -732,7 +732,7 @@ func (og *operationGenerator) GenerateUnmountVolumeFunc(
 	podsDir string) (volumetypes.GeneratedOperations, error) {
 
 	var pluginName string
-	if useCSIPlugin(og.volumePluginMgr, volumeToUnmount.VolumeSpec) {
+	if volumeToUnmount.VolumeSpec != nil && useCSIPlugin(og.volumePluginMgr, volumeToUnmount.VolumeSpec) {
```

Comment:
I think we should return an error in ConstructVolumeSpec() instead if we can't support reconstruction, like [here](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/volume/glusterfs/glusterfs.go#L232).

This will cause the Unmount() to happen, which will fix the filesystem-based paths, however block-based paths will not work because UnmountDevice() is still not called.

Reply:
Does raw block reconstruction have a similar issue? We don't have tests for raw block reconstruction: https://github.com/kubernetes/kubernetes/issues/74545

Reply:
cc @gnufied

Reply:
I discussed some more with @jingxu97, and she pointed out that even if we return error in ConstructVolumeSpec(), the volumeSpec will still be nil here, so we still need this check.

It looks like this check was introduced as part of CSI migration. 1) This should be feature gated, and 2) can `useCSIPlugin` return false if volume spec is nil? Does that make sense for all operations? cc @davidz627

Reply:
for 1) I've submitted a PR to make sure we do the migration feature gate check before doing anything else.

For 2) I think I need to do some more investigation and address it in the CSI Migration Reconstruction design.